### PR TITLE
Allow users to dislike ads from advertisers by clicking the ads history thumbs down button

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/17319

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Confirm disliking an ad stops ads from that advertiser

### CODE REVIEW

**Comment by tmancey** [comment]
Should be rebased after https://github.com/brave/brave-core/pull/10087 is merged and then the PR can be code reviewed

File: components/brave_ads/browser/ads_service_impl.cc [hunk 1]
```diff
@@ -1317,58 +1305,24 @@ void AdsServiceImpl::OnGetAdsHistory(OnGetAdsHistoryCallback callback,
...
+    base::ListValue ad_history_list;
+    ad_history_list.Append(std::move(ad_history_dictionary));
 
     base::DictionaryValue dictionary;
-
     dictionary.SetKey("uuid", base::Value(std::to_string(uuid++)));
-    auto time = base::Time::FromDoubleT(item.timestamp);
-    auto js_time = time.ToJsTime();
+    const base::Time time = base::Time::FromDoubleT(item.timestamp);
+    const double js_time = time.ToJsTime();
```
Comment (emerick) [inline components/brave_ads/browser/ads_service_impl.cc:1324]: Seems like the current recommendation is to use `ToJsTimeIgnoringNull`.

File: vendor/bat-native-ads/src/bat/ads/internal/client/preferences/filtered_advertiser_info.cc [hunk 2]
```diff
@@ -0,0 +1,51 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
```
Comment (emerick) [inline vendor/bat-native-ads/src/bat/ads/internal/client/preferences/filtered_advertiser_info.cc:1]: Nit: 2021?

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/client/preferences/filtered_advertiser_info.cc:1]: Fixed

File: components/brave_ads/browser/ads_service_impl.cc (hunk 1 above)
Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc:1324]: Fixed

**Review by emerick — APPROVED** [review]

File: vendor/bat-native-ads/src/bat/ads/category_content_info.cc [hunk 3]
```diff
@@ -25,6 +26,37 @@ bool CategoryContentInfo::operator!=(const CategoryContentInfo& rhs) const {
...
+  }
+
+  const std::string* category_value = dictionary->FindStringKey("category");
+  if (category_value) {
+    category = *category_value;
+  }
+
+  const absl::optional<int> opt_action_type_optional =
+      dictionary->FindIntKey("opt_action");
+  if (opt_action_type_optional && opt_action_type_optional.has_value()) {
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/category_content_info.cc:52]: this two conditions are checking the same:
`opt_action_type_optional ` and `opt_action_type_optional.has_value()`.  [Details here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/abseil-cpp/absl/types/optional.h;l=452?q=optional.h&ss=chromium)

could we remove one of them?

**Review by aseren — APPROVED** [review]

File: vendor/bat-native-ads/src/bat/ads/category_content_info.cc (hunk 3 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/category_content_info.cc]: Fixed

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/category_content_info.cc]: Fixed

**Pull request merged**